Speed-up GWR

Ziqi-Li opened the pull request:
Related to #4 
1. Move the fast fit feature from `GWR` class to `sel_bw` class. 
2. Revert `gwr.py` to very beginning since features in last PR are moved to 'sel_bw.py' where I think it could be.
3. Pass distance matrix in every kernel calculation to avoid repeated calculations.
4. Using OLS for Gaussian fit and IWLS for other Generalized models. This will make Gaussian fit faster, which may be the most common use case

ljwolf commented:
This is a pretty large change, we should probably tag more than one reviewer.

TaylorOshan commented:
Finally had a chance to take a close look at this. So, I get the logic of moving it from the GWR class to the bandwidth selection class. A few things I think we could work on in order to tighten up such a large refactor. 

(1) It looks as if none of the previous diagnostics are being used from the ``diagnostics.py`` module. I'm guessing that this is because they take a ``gwr`` object as input. One issue with defining the diagnostics in the ``sel_bw`` class is that then we have to maintain the math for those diagnostics in two different places. We could perhaps have a ``gwr_light`` object that could be instantiated with fewer attributes but could still have the necessary infrastructure to be able to be passed to diagnostics functions. That would mean we wouldn't need to have so much "fitting" code floating within the bandwidth selection routine. 

(2) Just noticed a double loop when computing spherical distance from lat/long. Would be possible to convert this to matrix form to avoid the loops? Could be prohibitive for large datasets.

(3) I'm wondering if we can also utilize the matrix algebra from the ``_compute_betas_gwr`` function in the ``iwls`` module so that the OLS routine used here is not duplicating code. IIRC, it is the same math for the OLS and GLM, just the GLM includes iterative weighting. So a single iteration of the iwls routine should give the same results as the OLS. Therefore we should be able to just use the ``_compute_betas_gwr`` function as is or use iwls with a flag that skips over the iterative weighting for the GLM routine. This may also be beneficial since the iwls routine is already designed to be able to use sparse matrices, which we mentioned might be useful in certain circumstances. 

(4) Keep in mind that new functions should also include new tests. 

Will take another look at this and see if I think of anything else.

Any thoughts, @ljwolf?

ljwolf commented on gwr/sel_bw.py:
```diff
@@ -207,24 +201,117 @@ def search(self, search='golden_section', criterion='AICc', bw_min=0.0,
+            
+    #hold it for allowing lat-lons in next PR
+    #_haversine formula to calculate distance
+    def _haversine(lat1, lon1, lat2, lon2):
```
this should not be defined in an internal method. Also, this should probably use the stuff in `libpysal.cg.sphere`.

Ziqi-Li replied:
the `_haversine() ` is not getting called in this PR, just a 'placeholder' for the next PRs which allowing spherical coords, I should have removed it since it is irrelevant. Yeah, I agree, it would be better to use `libpysal.cg.sphere` instead. Thanks for pointing that out.

ljwolf commented on gwr/sel_bw.py:
```diff
@@ -207,24 +201,117 @@ def search(self, search='golden_section', criterion='AICc', bw_min=0.0,
+    #_haversine formula to calculate distance
+    def _haversine(lat1, lon1, lat2, lon2):
+        R = 6371400 # Earth radius in meters
+        dLat = radians(lat2 - lat1)
```
Also, where is this `raidans` function coming from? it hasn't been imported and I don't see it defined?

ljwolf commented:
The `.pyc` functions need to be removed from the PR

ljwolf commented:
Just a few initial comments. I'll need a few days to read & understand.

Ziqi-Li commented:
Several updates regarding comments and suggestions from @ljwolf  and @TaylorOshan :
1. Removed `.pyc`
2. Removed `_haversine()` and perhaps using `libpysal.cg.sphere`  for handling lat-long.
3. Using `iwls._compute_betas_gwr()` to do solving instead of duplicating matrix inverse codes. The current `iwls()` for Gaussian requires two iterations since the betas are [initiated using zeros](https://github.com/pysal/spglm/blob/b225342e941faf828d5cb071fe705c4ca57c30b1/spglm/iwls.py#L48). Thus another (2nd) iteration is needed to converge. And thus it is twice slower than OLS. But directly calling `iwls._compute_betas_gwr()` only solves mat inverse once, which is faster.

TaylorOshan commented:
To ensure we move forward, we will take the simpler optimizations and get them factored in first and then we can discuss some of the more complicated optimizations after doing some profiling. Specifically, @Ziqi-Li and I discussed integrating the changes that thread the W matrix through the bandwidth selection so that it is not re-computed many times. In addition,  this PR will also introduce the optimization introduced in #15 . The optimization is straight forward, only involves the storage of a dictionary with single int/float key/value pairs, and is shouldn't be problematic with increasing ``n`` or ``p`` given the resolution of datasets we are currently able to feasibly analyze.  And then we will leave the refactor that minimizes the computation of diagnostics for another PR. The thinking is that after we integrate the W matrix optimization (and the previously introduced optimization of the golden_section routine where two unused computations are commented out) than we can do some profiling and see what the best way to minimize the computation time for diagnostics would be while also considering the use of sparse data structures.

ljwolf commented:
Great, sounds good to me

TaylorOshan commented on gwr/kernels.py:
```diff
@@ -61,11 +63,22 @@ def __init__(self, data, bandwidth=None, fixed=True, k=None,
             self.k = k
-        if points is None:
-            self.dmat = cdist(self.data, self.data)
+        
```
The logic here is kind of thick. After stepping it through it, it makes sense, so I don't see any issues for the PR, but we might think about a way simplify it in the future.

Specifically, I am referring to lines 66-81 and lines 101-111 where there is an interaction between whether or not the algorithm is "searching" for a bandwidth, carrying out a final GWR with the optimal bandwidth, or carrying out prediction, and whether or not an adaptive (needs dmat and sorted dmat) or fixed bandwidth (only needs dmat) is used.
